lib/Widgets/MessageDialog.vala

```diff
@@ -0,0 +1,242 @@
+                    break;
+                default:
+                    warning ("Unknown GtkButtonsType value");
+                    break;
+            }
+        }
+    }
+
+    /**
+     * The message grid that contains all the message widgets: the {@link Granite.MessageDialog.primary_label},
+     * {@link Granite.MessageDialog.secondary_label} and the image widget.
+     * 
+     * This grid can be used to add any custom widget to the message area such as a {@link Gtk.ComboBox} or {@link Gtk.CheckButton}.
+     * Note that after adding such widget you will need to call {@link Gtk.Widget.show} or {@link Gtk.Widget.show_all} on the widget itself for it to appear in the dialog.
+     */
+    public Gtk.Grid message_grid { get; construct; }
+
+    /**
+     * The image that's displayed in the dialog.
+     */
+    private Gtk.Image image;
+
+    /**
+     * Constructs a new {@link Granite.MessageDialog}.
+     * See {@link Gtk.Dialog} for more details.
+     * 
+     * @param primary_text the title of the dialog
+     * @param secondary_text the body of the dialog
+     * @param image_icon the {@link GLib.Icon} that is displayed on the left side of the dialog
+     * @param buttons the {@link Gtk.ButtonsType} value that decides what buttons to use, defaults to {@link Gtk.ButtonsType.CLOSE},
+     *        see {@link Granite.MessageDialog.buttons} on details and what values are accepted
+     */
+    public MessageDialog (string primary_text, string secondary_text, GLib.Icon image_icon, Gtk.ButtonsType buttons = Gtk.ButtonsType.CLOSE) {
+        Object (
+            resizable: false,
+            deletable: false,
+            skip_taskbar_hint: true,
+            transient_for: null,
+            primary_text: primary_text,
+            secondary_text: secondary_text,
+            image_icon: image_icon,
+            buttons: buttons
+        );
+    }
+
+    /**
+     * Constructs a new {@link Granite.MessageDialog} with an icon name as it's icon.
+     * This constructor is same as the main one but with a difference that 
+     * you can pass an icon name string instead of manually creating the {@link GLib.Icon}.
+     * 
+     * The {@link Granite.MessageDialog.image_icon} will store the created icon
+     * so you can retrieve it later with {@link GLib.Icon.to_string}.
+     * 
+     * See {@link Gtk.Dialog} for more details.
+     * 
+     * @param primary_text the title of the dialog
+     * @param secondary_text the body of the dialog
+     * @param icon_name the icon name to create the dialog image with
+     * @param buttons the {@link Gtk.ButtonsType} value that decides what buttons to use, defaults to {@link Gtk.ButtonsType.CLOSE},
+     *        see {@link Granite.MessageDialog.buttons} on details and what values are accepted
+     */
+    public MessageDialog.from_icon_name (string primary_text, string secondary_text, string icon_name = "dialog-information", Gtk.ButtonsType buttons = Gtk.ButtonsType.CLOSE) {
```

Comment:
Unclear from the name of the constructor what the icon is used for.

Reply:
@matzipan what name do you propose?

Reply:
Perhaps `with_image_from_icon_name`?